rapids-cmake/cpm/nvtx3.cmake

```diff
@@ -0,0 +1,107 @@
+rapids_cpm_nvtx3
+----------------
+
+.. versionadded:: v24.06.00
+
+Allow projects to find `NVTX3` via `CPM` with built-in tracking of dependencies
+for correct export support.
+
+Uses the version of nvtx3 :ref:`specified in the version file <cpm_versions>` for consistency
+across all RAPIDS projects.
+
+.. code-block:: cmake
+
+  rapids_cpm_nvtx3( [BUILD_EXPORT_SET <export-name>]
+                    [INSTALL_EXPORT_SET <export-name>]
+                    [<CPM_ARGS> ...])
+
+.. |PKG_NAME| replace:: nvtx3
+.. include:: common_package_args.txt
+
+Result Targets
+^^^^^^^^^^^^^^
+  nvtx3-c, nvtx3-cpp targets will be created
+
+Result Variables
+^^^^^^^^^^^^^^^^
+  :cmake:variable:`NVTX3_SOURCE_DIR` is set to the path to the source directory of nvtx3.
+  :cmake:variable:`NVTX3_BINARY_DIR` is set to the path to the build directory of nvtx3.
+  :cmake:variable:`NVTX3_ADDED`      is set to a true value if nvtx3 has not been added before.
+  :cmake:variable:`NVTX3_VERSION`    is set to the version of nvtx3 specified by the versions.json.
+
+#]=======================================================================]
+function(rapids_cpm_nvtx3)
+  list(APPEND CMAKE_MESSAGE_CONTEXT "rapids.cpm.nvtx3")
+
+  set(options)
+  set(one_value USE_PROPRIETARY_BINARY BUILD_EXPORT_SET INSTALL_EXPORT_SET)
+  set(multi_value)
+  cmake_parse_arguments(_RAPIDS "${options}" "${one_value}" "${multi_value}" ${ARGN})
+
+  # Fix up _RAPIDS_UNPARSED_ARGUMENTS to have EXPORT_SETS as this is need for rapids_cpm_find
+  if(_RAPIDS_INSTALL_EXPORT_SET)
+    list(APPEND _RAPIDS_EXPORT_ARGUMENTS INSTALL_EXPORT_SET ${_RAPIDS_INSTALL_EXPORT_SET})
+  endif()
+  if(_RAPIDS_BUILD_EXPORT_SET)
+    list(APPEND _RAPIDS_EXPORT_ARGUMENTS BUILD_EXPORT_SET ${_RAPIDS_BUILD_EXPORT_SET})
+  endif()
+  set(_RAPIDS_UNPARSED_ARGUMENTS ${_RAPIDS_EXPORT_ARGUMENTS})
+
+  set(to_install OFF)
+  if(INSTALL_EXPORT_SET IN_LIST ARGN)
+    set(to_install ON)
+  endif()
+
+  include("${rapids-cmake-dir}/cpm/detail/package_details.cmake")
+  rapids_cpm_package_details(nvtx3 version repository tag shallow exclude)
+
+  include("${rapids-cmake-dir}/cpm/detail/generate_patch_command.cmake")
+  rapids_cpm_generate_patch_command(nvtx3 ${version} patch_command)
+
+  include("${rapids-cmake-dir}/cpm/find.cmake")
+  rapids_cpm_find(NVTX3 ${version} ${ARGN}
+                  GLOBAL_TARGETS nvtx3-c nvtx3-cpp
+                  CPM_ARGS
+                  GIT_REPOSITORY ${repository}
+                  GIT_TAG ${tag}
+                  GIT_SHALLOW ${shallow} ${patch_command} SOURCE_SUBDIR c
+                  EXCLUDE_FROM_ALL ${exclude})
+
+  include("${rapids-cmake-dir}/cpm/detail/display_patch_status.cmake")
+  rapids_cpm_display_patch_status(NVTX3)
+
+  # Set up install rules
+  if(to_install)
+    include(GNUInstallDirs)
+    install(DIRECTORY "${NVTX3_SOURCE_DIR}/c/include/" DESTINATION "${CMAKE_INSTALL_INCLUDEDIR}")
+    install(TARGETS nvtx3-c nvtx3-cpp EXPORT ${_RAPIDS_BUILD_EXPORT_SET})
```

Comment:
What happens here if the user doesn't pass `_RAPIDS_BUILD_EXPORT_SET`?

Reply:
I don't really know the expected behavior here. I decided to try removing the "build export set" since we're not conditioning on that above. We were deciding whether to execute this block based on `if(INSTALL_EXPORT_SET IN_LIST ARGN)`. The build passed locally for me when I removed the "build export set" entirely. Take a look at 6ba6a376bb2184aec693097934dc8f71c37b8665 and see if it seems reasonable.

Reply:
The `BUILD_EXPORT_SET` should control the build directory exporting. If no INSTALL or BUILD export is requested that means `nvtx3` is a private dependency and doesn't need any exporting/installing.

Reply:
@robertmaynard Is any action required here, or is that comment just informational to answer Kyle's question?

Reply:
Kyle is point out that the condition of the branch is if `_RAPIDS_INSTALL_EXPORT_SET` has been specified, but you are calling the install rule with `_RAPIDS_BUILD_EXPORT_SET`. You can't presume that `_RAPIDS_BUILD_EXPORT_SET` is set in these cases, nor that it has the same value as `_RAPIDS_INSTALL_EXPORT_SET`

Reply:
I changed this in 6ba6a37 to remove usage of `_RAPIDS_BUILD_EXPORT_SET`. It appears to build without that. I don't know if it is needed.